File: crawler/src/models/product.py
from dataclasses import dataclass
from types import NoneType
from typing import List

from dataclasses_json import dataclass_json

from src.models.finn_ad import FinnAd


@dataclass_json
@dataclass(order=True, frozen=True)
class Category(object):
    main: int
    sub: int
    product: int


@dataclass_json
@dataclass(order=True, frozen=True)
class Retailer(object):
    name: str
    price: float
    sku: str
    url: str
    category: str

    @property
    def id(self):
        return f"{self.name}_{self.sku}"


# TODO: use pydantic
@dataclass_json
@dataclass(order=True)
class Product(object):
    name: str
    description: str
    image: str
    mpns: List[str]
    gtins: List[str]
    retailers: List[Retailer]
    finn_ads: List[FinnAd] = None
    brand: str = None
    category: Category | None = None
    finn_query: str | None = None
    id: int | None = None

    def __post_init__(self):
        if not self.finn_ads:
            self.finn_ads = []

        assert all(isinstance(mpn, str) for mpn in self.mpns)
        assert all(isinstance(gtin, str) for gtin in self.gtins)
        assert all(isinstance(retailer, Retailer) for retailer in self.retailers)
        assert all(isinstance(finn_ad, FinnAd) for finn_ad in self.finn_ads)
        assert isinstance(self.name, str)
        assert isinstance(self.description, str)
        assert isinstance(self.image, str)
        assert isinstance(self.brand, (NoneType, str))
        assert isinstance(self.category, (NoneType, Category))
        assert isinstance(self.finn_query, (NoneType, str))
        assert isinstance(self.id, (NoneType, int))

File: website/src/models/product.ts
import {
  Product as PrismaProduct,
  Gtin,
  Mpn,
  ProductRetailer,
} from "@prisma/client";
import { JsonObject } from "@prisma/client/runtime/library";

export interface Product {
  id: number;
  name: string;
  finn_query: string | null;
  brand: string | null;
  description: string;
  image: string;
  mpns: string[];
  gtins: string[];
  retailers: Retailer[];
  category: Category | null;
}

export interface Category {
  main: number;
  sub: number;
  product: number;
}

export interface Retailer {
  name: string;
  price: number;
  sku: string;
  url: string;
  category: string;
}

export function fromPrismaProduct(
  p: PrismaProduct & {
    gtins: Gtin[];
    mpns: Mpn[];
    retailers: ProductRetailer[];
  }
) {
  return {
    id: p.id,
    name: p.name,
    finn_query: p.finn_query,
    brand: p.brand,
    description: p.description,
    gtins: p.gtins.map((g) => g.gtin),
    mpns: p.mpns.map((m) => m.mpn),
    image: p.image,
    retailers: p.retailers.map((r) => ({
      category: r.category,
      name: r.name,
      price: r.price,
      sku: r.sku,
      url: r.url,
    })),
    category: p.category
      ? {
          main: (p.category as JsonObject).main as number,
          sub: (p.category as JsonObject).sub as number,
          product: (p.category as JsonObject).product as number,
        }
      : null,
  };
}
